homeassistant/components/bluetooth/models.py

```diff
@@ -0,0 +1,146 @@
+        return
+
+    if (uuids := filters.get(FILTER_UUIDS)) and not uuids.intersection(
+        advertisement_data.service_uuids
+    ):
+        return
+
+    try:
+        callback(device, advertisement_data)
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.exception("Error in callback: %s", callback)
+
+
+class HaBleakScanner(BleakScanner):  # type: ignore[misc]
+    """BleakScanner that cannot be stopped."""
+
+    # HaBleakScanner is a singleton
+    _callbacks: list[tuple[AdvertisementDataCallback, dict[str, set[str]]]] = []
+    _history: LRU = LRU(MAX_HISTORY_SIZE)
+
+    @hass_callback
+    def async_register_callback(
+        self, callback: AdvertisementDataCallback, filters: dict[str, set[str]]
+    ) -> CALLBACK_TYPE:
+        """Register a callback."""
+        callback_entry = (callback, filters)
+        self._callbacks.append(callback_entry)
+
+        @hass_callback
+        def _remove_callback() -> None:
+            self._callbacks.remove(callback_entry)
+
+        # Replay the history since otherwise we miss devices
+        # that were already discovered before the callback was registered
+        # or we are in passive mode
+        for device, advertisement_data in self._history.values():
+            _dispatch_callback(callback, filters, device, advertisement_data)
+
+        return _remove_callback
+
+    def async_callback_dispatcher(
+        self, device: BLEDevice, advertisement_data: AdvertisementData
+    ) -> None:
+        """Dispatch the callback.
+
+        Here we get the actual callback from bleak and dispatch
+        it to all the wrapped HaBleakScannerWrapper classes
+        """
+        self._history[device.address] = (device, advertisement_data)
+
+        removes = []
+
+        for callback_filters in self._callbacks:
+            if callback_filters[0] is None:
+                removes.append(callback_filters)
+                continue
+            _dispatch_callback(*callback_filters, device, advertisement_data)
+
+        # If the callback is None, it means it was destroyed
+        # so we need to remove it from the list
+        for callback_filters in removes:
+            self._callbacks.remove(callback_filters)
+
+
+class HaBleakScannerWrapper(BleakScanner):  # type: ignore[misc]
+    """A wrapper that uses the single instance."""
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initialize the BleakScanner."""
+        self._detection_cancel: CALLBACK_TYPE | None = None
+        self._filters: dict[str, set[str]] = {}
+        if "filters" in kwargs:
+            self._mapped_filters = {k: set(v) for k, v in kwargs["filters"].items()}
+        if "service_uuids" in kwargs:
+            self._mapped_filters[FILTER_UUIDS] = set(kwargs["service_uuids"])
+        super().__init__(*args, **kwargs)
+
+    async def stop(self, *args: Any, **kwargs: Any) -> None:
+        """Stop scanning for devices."""
+        return
```

Comment:
Should we raise so we find out what processes try to stop us ?

Reply:
Its normal for them to call stop so probably not